packages/firebase_core/firebase_core_platform_interface/lib/firebase_core_platform_interface.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:meta/meta.dart' show visibleForTesting;
+
+import 'src/firebase_app_data.dart';
+import 'src/firebase_options.dart';
+import 'src/method_channel_firebase_core.dart';
+
+export 'src/firebase_app_data.dart';
+export 'src/firebase_options.dart';
+export 'src/method_channel_firebase_core.dart';
+
+/// The interface that implementations of `firebase_core` must extend.
+///
+/// Platform implementations should extend this class rather than implement it
+/// as `firebase_core` does not consider newly added methods to be breaking
+/// changes. Extending this class (using `extends`) ensures that the subclass
+/// will get the default implementation, while platform implementations that
+/// `implements` this interface will be broken by newly added
+/// [FirebaseCorePlatform] methods.
+abstract class FirebaseCorePlatform {
+  /// Only mock implementations should set this to `true`.
+  ///
+  /// Mockito mocks implement this class with `implements` which is forbidden
+  /// (see class docs). This property provides a backdoor for mocks to skip the
+  /// verification that the class isn't implemented with `implements`.
+  @visibleForTesting
+  bool get isMock => false;
+
+  /// The default instance of [FirebaseCorePlatform] to use.
+  ///
+  /// Platform-specific plugins should override this with their own class
+  /// that extends [FirebaseCorePlatform] when they register themselves.
+  ///
+  /// Defaults to [MethodChannelFirebaseCore].
+  static FirebaseCorePlatform get instance => _instance;
+
+  static FirebaseCorePlatform _instance = MethodChannelFirebaseCore();
+
+  // TODO(amirh): Extract common platform interface logic.
+  // https://github.com/flutter/flutter/issues/43368
+  static set instance(FirebaseCorePlatform instance) {
+    if (!instance.isMock) {
```

Comment:
I need this on my platform interfaces, but I'm not the biggest fan of this runtime check. 

I understand this may be late, and probably beating on a dead horse, but why not something like [this](https://dartpad.dartlang.org/c8095fa64d7ab880765861d74a74aa05)?

It satisfies the "do not break on unimplemented methods" and doesn't need a runtime check. Too much friction with the `extends Plugin implements WhateverPlatform`? What is missing that I'm not seeing?

Reply:
I prefer the current approach to the `noSuchMethod` approach. `noSuchMethod` can actually cause ahead-of-time compilers to generate much slower and bigger code. Stephen from the dart2js team gave a good talk a few years ago on why `noSuchMethod` is harmful for `dart2js`: https://youtu.be/xneWkyKLuk0?t=1107

Reply:
I knew I was missing something, thanks for the link!